wallet-new/test/state-machine/Wallet.hs

```diff
@@ -142,98 +156,155 @@ preconditions _ ResetWalletA      = Top
-            updatedWallets = map (first update) $ filter (thisWallet . fst) mWallets
-            update w = w { V1.walAssuranceLevel = uwalAssuranceLevel, V1.walName = uwalName }
-        in model { mWallets = updatedWallets <> filter (not . thisWallet . fst) mWallets }
-    UpdateWalletPasswordA wId V1.PasswordUpdate{..} ->
-        let thisWallet (wal, pass) = wId == V1.walId wal && pass == Just pwdOld
-            updatedWallets = map (\(wal,_) -> (wal, Just pwdNew)) $ filter thisWallet mWallets
-        in model { mWallets = updatedWallets <> filter (not . thisWallet) mWallets }
-    DeleteWalletA wId ->
-        let thisWallet = (wId ==) . V1.walId
-        in model { mWallets = filter (not . thisWallet . fst) mWallets }
+transitions model@Model{..} cmd res = case (cmd, res) of
+    (ResetWalletA, ResetWalletR) -> Model mempty mempty 0 True
+    (ResetWalletA, _) -> shouldNotBeReachedError
+    (CreateWalletA (WL.CreateWallet V1.NewWallet{..}), CreateWalletR (Right wallet)) ->
+        -- TODO: use lenses
+        model { mWallets = (wallet, newwalSpendingPassword) : mWallets }
+    (CreateWalletA _, CreateWalletR (Left _)) -> increaseUnhappyPath
+    (CreateWalletA _, _) -> shouldNotBeReachedError
+    (GetWalletsA, GetWalletsR _) -> model
+    (GetWalletsA, _) -> shouldNotBeReachedError
+    (GetWalletA _, GetWalletR (Right _)) -> model
+    (GetWalletA _, GetWalletR (Left _)) -> increaseUnhappyPath
+    (GetWalletA _, _) -> shouldNotBeReachedError
+    (UpdateWalletA wId _, UpdateWalletR (Right wallet)) ->
+        let thisWallet = (wId ==) . V1.walId . fst
+        in model { mWallets = filter thisWallet mWallets <> filter (not . thisWallet) mWallets }
+    (UpdateWalletA _ _, UpdateWalletR (Left _)) -> increaseUnhappyPath
+    (UpdateWalletA _ _, _) -> shouldNotBeReachedError
+    (UpdateWalletPasswordA wId V1.PasswordUpdate{..}, UpdateWalletPasswordR (Right wallet)) ->
+        let thisWallet = (wId ==) . V1.walId . fst
+        in model { mWallets = (wallet, Just pwdNew) : filter (not . thisWallet) mWallets }
+    (UpdateWalletPasswordA _ _, UpdateWalletPasswordR (Left _)) -> increaseUnhappyPath
+    (UpdateWalletPasswordA _ _, _) -> shouldNotBeReachedError
+    (DeleteWalletA wId, DeleteWalletR (Right _)) ->
+        let thisWallet = (wId ==) . V1.walId . fst
+        in model { mWallets = filter (not . thisWallet) mWallets }
+    (DeleteWalletA _, DeleteWalletR (Left _)) -> increaseUnhappyPath
+    (DeleteWalletA _, _) -> shouldNotBeReachedError
+    (CreateAccountA _ _, CreateAccountR (Right account)) ->
+        model { mAccounts = account : mAccounts }
+    (CreateAccountA _ _, CreateAccountR (Left _)) -> increaseUnhappyPath
+    (CreateAccountA _ _, _) -> shouldNotBeReachedError
+
+    -- NOTE: don't use catch all pattern like
+    --
+    -- (_, _) ->
+    --
+    -- as it will most likely bite us.
+  where
+    -- TODO: use postcondition that ration of unhappy paths has to be expected (ie, similar to test coverage)
```

Comment:
Interesting approach with the unhappy path counter... You could perhaps use the `invariant` field in the `StateMachine` record to enforce some threshold (`invariant` gets checked after every action).

An alternative way, which maybe you've already thought about, is to add a bad variant for each action that can go wrong. Something like:

```haskell
CreateAccountA ...
CreateAccountBadA ...
```
The precondition for `CreateAccountBadA` should be the negation of the precondition of `CreateAccountA`, then you can have the post-condition for `CreateAccountA` always enforce a `Right` response while the post-condition for `CreateAccountBadA` always enforce a `Left` response.

The second approach will be more code, but also give you more control and a more fine grained specification, as you can assert exactly what errors should occur when some specific preconditions are violated.

Reply:
Thank you for prompt review!

> Interesting approach with the unhappy path counter... You could perhaps use the invariant field in the StateMachine record to enforce some threshold (invariant gets checked after every action).

I will have a look at `invariant` field (probably next week) - I have noticed it but forgot about that.

> An alternative way, which maybe you've already thought about, is to add a bad variant for each action that can go wrong. Something like ...

yes this was on my mind. To generate bad actions (that would eventually trigger error and/or `Left` branches) to increase the coverage.

This week I am trying to cover most of happy paths (`Right` branches) but keeping in mind and preparing the ground for the rest.

Thanks <3